### ISSUE TITLE
In 0.9.1

### ISSUE THREAD

**Issue by felipejfc** [Automattic/kue#636]
I'm getting stucked jobs in queued state...
sympton is: a lot of jobs fail with TTL, then, after returning to the queue only 1 job at a time is getting promoted to active, however my queue is configured to have 100 jobs being processed concurrently...

**Comment by behrad** [comment]
since your jobs are failing with TTL, it shows your workers are missing to call `done` and somehow stuck in their process function. 
They will be stuck until you restart the node process.

If this is the case, it is your app's fault. Your process functions should prevent from being stuck.
Kue's fault is not to open up the stuck worker when a job's TTL is fired.

We should dig more into details to see if I'm true... can you provide a sample code?

**Comment by felipejfc** [comment]
Indeed it is because I'm not calling done(), and the workflow is as expected, I do not call done and then the job fails because of ttl, so far so good, problem is I after that this somehow messes with the job processing, I cannot provide the code, however I can tell what I'm doing, it is a HTML Request and some scrapping. I suspect that the request is never completing so I'm stuck and never call done, then the job fail with ttl as it was supposed to... however, after that in some ocasions the jobs never process of anymore, only if I restart node process...

**Comment by behrad** [comment]
I should check if Kue is correctly opening up the worker whose job is TTL-d or not :)

**Comment by jamc** [comment]
Im experiencing a similar issue, after days of excellent working, suddenly it starts to give TTL Exceeded 70 to 80% of the times.

My workers start logging from the 1 line but the only logs I have is enqueueing the job and the TTL Exceeded handler, no worker processing.

On Kue dashboards, the Job passes from Queued to Active, then Delayed, Active and so until it reaches the maximum tries and gives ups, all tries fails with TTL Exceeded.

So far to fix this we restart the instances an clear the redis database.

An interesting thing is that the hgetall that Kue uses is taking 6 seconds to complete, Im suspecting that redis is being drowned but i dont know how because we have pretty large servers.

Another piece of information, we are reaching the 4000 processed jobs.

I you have any clue or guide please let me know, we face a total customer impact and production getting stuck every time this happens.

Thanks!

**Comment by behrad** [comment]
which version of Kue and redis are you on @jamc ?

from your explanation I'm not sure your issue is the same as @felipejfc 


> An interesting thing is that the hgetall that Kue uses is taking 6 seconds to complete, Im suspecting that redis is being drowned but i dont know how because we have pretty large servers.

Can you confirm this? I suspect the internal promote & ttl scripts. It is so hard to drown redis by normal commands at your rate (4000 jobs...) this may be the cause that your jobs are waited and ttld!? how much your `ttl` is?

**Comment by jamc** [comment]
@behrad i checked the enqueuing and workers and both have version 0.9.3.

I have a TTL of 30 seconds and 3 retries.

This is the base code i use to enqueue
```javascript
createJob.create('createJob', {
  title: 'create resource'
  // ... more variables here
})
.attempts(3)
.backoff( {delay: 500, type:'exponential'} )
.ttl(30000)
.removeOnComplete(true)
.save(function(err) {
  if (err) {
    throw err;
  }
  // ... more code here
});
```

And this is the one i use to process:
```javascript
var jobs = kue.createQueue({
  prefix: 'myPrefix',
  redis: {
    createClientFactory: function(){
      return myRedisInstance;
    }
  }
});

if (cluster.isMaster) {
  kue.app.use(someExpressMiddleware);
  kue.app.listen(3000);
  jobs.promote();
  function handleExitEvent(code, signal) {
    cluster.fork().on('exit', handleExitEvent);
  }
  for (var i = 0; i < 3; i++) {
    cluster.fork().on('exit', handleExitEvent);
  }

  jobs.on('job failed', function(id, errorMessage){
    if (errorMessage === 'TTL exceeded') {
      kue.Job.get(id, function( err, job ) {
        // Clean up failed job
      });
    }
  });

} else {
  jobs.process('create', 5, function(job, done) {
    functionThatProcess(job, done, moreVars);
    })();
  });
}
```

Thanks!

**Comment by behrad** [comment]
Can i see `handleExitEvent` ?
depends whats exactly happening in your `functionThatProcess`.
I don't think I can generate this with the code you provided as this is a very basic Kue code that should work. Can you test this on your server to confirm me?

**Comment by jamc** [comment]
Sure:

```javascript
function handleExitEvent(code, signal) {
  cluster.fork().on('exit', handleExitEvent);
}
```

I think that `functionThatProcess` code does not matter due that I doesn't get there, I can affirm that because I don't see any logging from that function. It seems that the worker can't get the Job.

I have more information, the number of connections opened to redis was incredible high, 15000 open connections, we dropped all connections and it started working again.

It looks like node is getting obfuscated for that number of opened connections, the thing here is that I have 8 apps running that use redis and not all use Kue, so i can not ensure that Kue is not closing connections but can be a lead.

**Comment by behrad** [comment]
Kue opens fixed number of connections to redis, and reuses them throughout its life. 15000 redis connections is something related to other apps. I think you should first resolve that issue then.

**Comment by jamc** [comment]
Yes, I agree with that, im going through my apps to test the connections, I'll keep you informed, thanks for the support.

**Comment by sorin** [comment]
We have the same issue, its pretty easy to replicate:

- create job with concurrency:1 and ttl:X seconds
- enqueue 2 jobs that time out (never call done())
- job 1 becomes active, X seconds later it moves to failed with "TTL exceeded" as expected, but job 2 never becomes active
- restarting kue works, job 2 now becomes active

Seems to me like this is a serious problem that pretty much renders TTL functionality useless.

**Comment by sorin** [comment]
I tried replicating the problem with in unit tests and couldnt. Going to do some more investigations.
Sent a PR with the tests regardless, I think they're valuable: https://github.com/Automattic/kue/pull/720

**Comment by behrad** [comment]
will check this thank you...

**Comment by Paveltarno** [comment]
Any news on this? I'm getting this on 0.10.5
